model/views: Add settings support for "playground" feature (backend)

sumanthvrao opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR adds the database model for configuring realm playground settings and the API endpoints for `POST` and `DELETE`. 

**Testing Plan:** <!-- How have you tested? -->
Backend tests added and manual testing done.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

### TODO
(Things mentioned in #16432 and more)

- [ ] Build a configuration UI (like the Linkifiers UI).
- [ ] Figure out how to handle the default playgrounds for an organization.
- [ ] Investigate what services like `repl.it` exist and contact a few for permission when thinking about what defaults to support.
- [ ] Document disabling the feature for organizations who don't want a button that sends code to a third party.
- [ ] Document the feature on /features, /for/open-source, /help/format-...-markdown, etc and link to it from within the open-api documentation

With this PR, the backend part of #16432 would be completed.

sumanthvrao commented:
@timabbott Opened up this WIP PR for some early feedback :)

timabbott commented on zerver/lib/export.py:
```diff
@@ -626,6 +628,13 @@ def get_realm_config() -> Config:
+        model=RealmPlayground,
+        normal_parent=realm_config,
+        parent_key='realm_id__in',
+    )
```
I can't remember -- do we need code on the `zerver/lib/import_realm.py` side as well?

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5551,6 +5552,14 @@ def do_remove_realm_domain(realm_domain: RealmDomain, acting_user: Optional[User
+    realm_playground = RealmPlayground(realm=realm, **playground_info)
+    realm_playground.full_clean()
+    realm_playground.save()
+    event = dict(type="realm_playground", op="add", realm_playground=playground_info)
```
We may want to have these events just send the whole data set, rather than doing an add/remove system, because it's expected to be a consistently small data set, and that makes the client-side code simpler.

timabbott commented on zerver/tests/test_realm_playgrounds.py:
```diff
@@ -0,0 +1,88 @@
+        payload = {
+            "name": "Python playground",
+            "pygments_language": "Python",
+            "url_prefix": "https://some-website.com/?code="
```
Let's use E.g. `python.example.com` for these examples -- that's the RFC example domain.

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -0,0 +1,27 @@
+@has_request_variables
+def add_realm_playground(request: HttpRequest, user_profile: UserProfile,
+                         name: str=REQ(), pygments_language: str=REQ(),
+                         url_prefix: str=REQ()) -> HttpResponse:
```
We generally prefer to do validation here, in the REQ parameters, rather than via a try/except with model fields (even if the latter is useful secondary validation to ensure we don't save anything problematic to the database)

sumanthvrao replied:
Quick doubt here, I know you mentioned doing the validation inside REQ parameters,  but for the regex and url cases, we want to do some custom validation (beyond those offered by `lib/validators`. I'm thinking of adding check functions like this in the view with the functions themselves in `lib/playground.py`. 
```
def add_realm_playground(request: HttpRequest, user_profile: UserProfile,
                         name: str=REQ(), pygments_language: str=REQ(),
                         url_prefix: str=REQ()) -> HttpResponse:
    check_valid_pygments_language(pygments_language)
    check_valid_playground_url(url_prefix)
```

timabbott replied:
Yeah, that's the way to do it if `validators.py` doesn't support the validation we need.

(Well, or we could write a custom validator in the style of `validators.py`).

timabbott commented on zproject/urls.py:
```diff
@@ -287,6 +288,11 @@
+    # realm/playgrounds -> zerver.views.realm_playgrounds
+    rest_path('realm/playgrounds',
+              POST=(add_realm_playground,
+                    {'intentionally_undocumented'})),
```
Rather than tagging this as intentionally_undocumented, add it to the exclude list in `test_openapi.py`; we do intend to document endpoints like this one.

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -25,3 +26,15 @@ def add_realm_playground(request: HttpRequest, user_profile: UserProfile,
+    except RealmPlayground.DoesNotExist:
+        return json_error(_("Playground does not exist"))
+
+    playground_object.delete()
```
This should be a `do_*` method that triggers a `send_event()` call to update clients.

Also be sure to add a `test_events.py` test for both the add and delete code paths.

sumanthvrao commented:
@timabbott Pushed a bunch of changes here. It's ready for another round of review.

A couple of points before reviewing:
* I'm not 100% sure of the import_realm changes above and you may want to review that in more detail.
* Do let me know if there is a better way to organize what change goes into which commit/split into further commits.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -6500,6 +6501,13 @@ def do_remove_realm_domain(
     send_event(realm, event, active_user_ids(realm.id))
 
 
+def do_add_realm_playground(realm: Realm, playground_info: Dict[str, str]) -> int:
```
Probably we should call this `check_add_realm_playground` -- we try to reserve `do_` names for things where all input validation is done and we don't expect failures.

timabbott replied:
Oh, I think it's OK but we should add a comment that we expect `full_clean` to always pass because `check_url` does the same validation above.

timabbott commented on zerver/tests/test_realm_playgrounds.py:
```diff
@@ -0,0 +1,80 @@
+            "url_prefix": "https://invalid-url",
+        }
+        resp = self.api_post(iago, "/json/realm/playgrounds", self.json_serialize(payload))
+        self.assert_json_error(resp, "url_prefix is not a URL")
```
I think this has test coverage for your coverage comment (which we try to avoid wherever possible).

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -0,0 +1,44 @@
+        playground_id = do_add_realm_playground(
+            realm=user_profile.realm, playground_info=playground_dict
+        )
+        return json_success({"id": playground_id})
```
I would probably put this after the `ValidationError` exception block, mainly from the philosophy that exception blocks should be as small as possible.

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -0,0 +1,44 @@
+            realm=user_profile.realm, playground_info=playground_dict
+        )
+        return json_success({"id": playground_id})
+    except ValidationError as e:  # nocoverage
```
We should remove this nocoverage comment; I think you actually test this code path already anyway.

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -0,0 +1,44 @@
+        "name": name.strip(),
+        "pygments_language": pygments_language.strip(),
+        "url_prefix": url_prefix.strip(),
+    }
```
Personally I'd have just made these either positional parameters or required kwargs to `do_add_realm_playground` -- that way the type-checker does more useful validation for you.

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -0,0 +1,44 @@
+
+def check_pygments_language(var_name: str, val: object) -> str:
+    s = check_string(var_name, val)
+    valid_pygments_language = re.compile(r"^[ a-zA-Z0-9_+-./#]*$")
```
Do we want to do this sort of regular expression check, or instead try to check it's actually in the list of pygments languages that 

I feel like we talked about this before, so probably this is just a case where we should write down the reasoning in a comment here.

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -42,3 +42,18 @@ def add_realm_playground(
+    request: HttpRequest, user_profile: UserProfile, playground_id: int
+) -> HttpResponse:
+    try:
+        RealmPlayground.objects.get(id=playground_id)
```
This is a security bug waiting to happen, since we split the access from the validation that it'
s the same realm.  What we should do is have an `access_playground(user_profile, playground_id)` that fetches the object, filtering by realm as well, and returns an "Invalid playground" error for any ID not present in the current realm.  And then we can just pass the RealmPlayground object to `do_remove_realm_playground`.

For more complicated features, we'd put that access function in another file, but it's fine to just put it at the top of this file instead since we don't have a larger `zerver/lib/realm_playground.py`.

timabbott commented on zerver/models.py:
```diff
@@ -966,6 +966,19 @@ def __str__(self) -> str:
+                url_prefix=playground.url_prefix,
+            )
+        )
+    return playgrounds
```
I'd squash this into the first commit using it -- we don't like adding a function in a commit that doesn't have tests for it, since it's incomplete in that sense.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9170,6 +9211,25 @@ components:
+        name:
+          type: string
+          description: |
+            The name of the playground.
```
I'd expand this slightly more to explain this is how users will be able to select which playground to use if there are multiple options.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -1326,6 +1329,20 @@ def test_realm_domain_events(self) -> None:
+        )
+        check_realm_playgrounds("events[0]", events[0])
+
+        events = self.verify_action(lambda: do_remove_realm_playground(self.user_profile.realm, 1))
```
Let's not hardcode 1 here -- that'll break if we adjust `populate_db` to configure some (which we probably will want to do later in this feature's development, to make manual testing more convenient).  Instead, we should save the ID from the previous step.  (Or `last_id = RealmPlayground.objects.last().id`, if that's annoying with the lambda structure)

timabbott commented:
OK, I just posted a batch of comments that cover I think everything that'll need to change for the backend for this feature to be merged.  A few other notes:
* Can you open issues with zulip-terminal and zulip-mobile for this feature, with the "api migrations" label?  That'll give them a heads-up that this is a feature they may want to integrate.  (Probably not a high near-term priority for either).

Sorry for the slow review on this!  I think we probably only need one more round trip on this PR.  (Probably worth just doing a new PR for the frontend, since this is a pretty complete self-contained backend PR).

timabbott commented on zerver/models.py:
```diff
@@ -931,6 +931,54 @@ def flush_per_request_caches() -> None:
+
+    # User-visible display name used when configuring playgrounds in the settings page and
+    # displaying them in the playground links popover.
+    name: str = models.CharField(db_index=True, max_length=MAX_PLAYGROUND_NAME_LENGTH)
```
This mostly doesn't matter since this table will be tiny, but we probably want to use `TextField` for this -- in that most names will be much shorter than the max length, and so it's more efficient space-wise (and if we change our mind about the length) to use that structure instead.

timabbott commented on zerver/views/realm_playgrounds.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def check_pygments_language(var_name: str, val: object) -> str:
+    s = check_string(var_name, val)
```
We should use e.g. `check_capped_string` (I think) -- the ones that check the length.  In general, we want to do as much validation as possible close to the client.

(FWIW we generally avoid Django's validators system, in favor of doing all input validation in `REQ`)

timabbott commented:
It looks like CI is failing in the OpenAPI tests. 

As a sidenote, I think some of the OpenAPI documentation changes in the last commit belong in the previous couple commits (the ones adding new API endpoints); also the final commit should update `changelog.md` with a few API_FEATURE_LEVEL etc.

sumanthvrao commented:
@timabbott Thanks for giving this a detailed review!  I have pushed an update that addresses each of the comments you left. Find a changelog of this version compared to the previous, below:

- Renumbered the migration
- Added the API documentation for both the endpoints (along with python & curl examples); I caught a small typo when reading the "Documenting REST API" guide and fixed it in the first commit :) 
- The RealmPlayground model name is now a TextField `name: str = models.TextField(db_index=True)`
- Added a comment to the `do_add_realm_playground` function on how we handle input validation.
- I removed the no-coverage comment. We in fact didn't have any test which covered the ValidationError codepath in the view, so I added a test case for this (see `RealmPlaygroundTests.test_create_already_existing_playground` ).
- `check_string` is now `check_capped_string` which uses `RealmPlayground.MAX_PYGMENTS_LANGUAGE_LENGTH`
- Added a comment about not restricting the language to be from the list of Pygments languages.
- Moved the `json_success` after the `ValidationError`
- The "Add a playground" codepath now passes the parameters as a kwargs to its `do_*` function.
- Thanks for catching the security pitfall! I have changed delete playground workflow to first do an `access_playground_by_id` check and fetche the realm playground object, later passing it to the `do_remove_*` function.
- I removed hardcoding `id=1` and instead using `last_id` in the events tests, based on what you suggested.
- Lastly, I updated the changelog and bumped the API_FEATURE_LEVEL.

(btw I'll opening up issues in the terminal and mobile repos as you mentioned and will link to this from them)

timabbott commented:
Merged, with the following changes:
* Added an `id` field to the RealmPlayground dictionaries -- we need to include that field so that clients have the ID to pass to the `DELETE` endpoint.  (And made appropriate documentation changes).
* A few small edits to the API documentation.  We should make a note to add links from the API documentation to the Help Center to explain what this feature is once we have the UI finished and thus Help Center docs for it.

Thanks @sumanthvrao!  It'll be nice to have this API change finished.  I think the next step is hooking things up so that playgrounds configured using this system are actually used on the frontend, and then the final step will be adding the "organization settings" UI for configuring them, which can hopefully crib a bit off the linkifiers page (though that one is certainly not a paragon of quality; see https://github.com/zulip/zulip/pull/17045 and https://github.com/zulip/zulip/pull/17596).

We'll need to be a bit careful in doing the frontend work to avoid breaking the hardcoded ones we have; I think a good theory for v1 is to just use the hardcoded set if the `realm_playground` dict is empty, and then plan to remove that code in the commit that makes actual UI for creating these.
